### DISCUSSION TITLE
One question of  Openglobus

### DISCUSSION THREAD

**Discussion by ywszxdd** [openglobus/openglobus#590] [General]
I want to  render custom data(vertex in lonlat、colors、incides、etc) using openglobus as one layer，How shuold I do?

**Comment by Zemledelec** [comment]
Hi @ywszxdd! Do you want some colors on the surface?

**Reply by ywszxdd** [reply]
for example,i want to custom one triangle polygon in openglobus ,each of the triangle ‘s vertex has different color, like use webgl render one triangle ,and the triangle colud as one layer in the scene, if use Cesium,I could use Cesium.Geometry do it ,like this,How should I do in openglobus?

**Reply by ywszxdd** [reply]
Hi@Zemledelec?Could you give me a reply?

**Comment by ywszxdd** [comment]
and the triangle's vertex  has different lon、lat、height

**Reply by ywszxdd** [reply]
I use GeoObject,but GeoObject only support one color for one GeoObject . And then I use handler.gl  custom render it,but the render result's position cannot change when the camera or the earth change

**Comment by ywszxdd** [comment]
this is my code,in this ,the triangle could render,but cannot change when the camera or the earth change.
<script id="shader-vs" type="x-shader/x-vertes">
        attribute vec3 a_position;
        void main(){
          gl_Position = vec4(a_position,1);
        }
    </script>
    <script id="shader-fs" type="x-shader/x-fragment">
        precision mediump float;
        void main(){
          gl_FragColor = vec4(1,0,0.5,1);
        }
    </script>
    <script type="module">
        'use strict';
        import { Globe } from '../../src/og/Globe.js';
        import { XYZ } from '../../src/og/layer/XYZ.js';
        import { GeoTexture2d } from '../../src/og/layer/GeoTexture2d.js';
        import { GlobusTerrain } from '../../src/og/terrain/GlobusTerrain.js';
        import { Framebuffer } from '../../src/og/webgl/Framebuffer.js';
        import { Program } from '../../src/og/webgl/Program.js';
        import { wgs84 } from '../../src/og/ellipsoid/wgs84.js';
        let osm = new XYZ("OpenStreetMap", {
            isBaseLayer: true,
            url: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
            visibility: true,
            attribution: 'Data @ OpenStreetMap contributors, ODbL'
        });
        let globus = new Globe({
            "target": "globus",
            "name": "Earth",
            "terrain": new GlobusTerrain(),
            "layers": [osm]
        });
        let handler = globus.renderer.handler;
        handler.addProgram(new Program("vectorMap", {
            attributes: {
                'a_position': 'vec4'
            },
            vertexShader: document.getElementById("shader-vs").innerHTML,
            fragmentShader: document.getElementById("shader-fs").innerHTML
        }));
        var p1 = wgs84.lonLatToCartesian({ lon: 0, lat: 0, height: 10000 }).normal();
        var p2 = wgs84.lonLatToCartesian({ lon: 100, lat: 0, height: 5000 }).normal();
        var p3 = wgs84.lonLatToCartesian({ lon: 50, lat: 50, height: 1000 }).normal();
        var positions = [
            p1.x, p1.y, p1.z,
            p2.x, p2.y, p2.z,
            p3.x, p3.y, p3.z,
        ]
        var vertBuffer = r.handler.createArrayBuffer(new Float32Array(positions), 3, 3);
        globus.renderer.events.on("draw", function (r) {
            let gl = r.handler.gl;
            gl.clearColor(0.0, 0.0, 0.0, 1.0);
            gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

            r.handler.programs.vectorMap
                .set({
                    'a_position': vertBuffer,
                })
                .drawArrays(r.handler.gl.TRIANGLES, vertBuffer.numItems);
        });
    </script>

**Reply by pavletto** [reply]
I think it's because you don't use projection and view matrix. Don't understend why you write all newly. We have ready solutions. About geoObjects - you can extended `GeoObjectHandler` by method which can take an array of values of different colors for shape.

**Reply by pavletto** [reply]
also i'm in progress with textures for GeoObjects

**Reply by ywszxdd** [reply]
> I think it's because you don't use projection and view matrix. Don't understend why you write all newly. We have ready solutions. About geoObjects - you can extended `GeoObjectHandler` by method which can take an array of values of different colors for shape.

thank you very much,could you give me one example code about   extending GeoObjectHandler by method which can take an array of values of different colors for shape.

**Reply by pavletto** [reply]
Let me first find out why you need such a function. It's just not a typical use of geo objects. They are more suitable for uv texturing or solid color objects than for gradient color blending.

**Reply by ywszxdd** [reply]
I need render the 3D radar echo,20 levels,each level has 500*500 grid polygon,each polygon has its own color.  And render the 3D radar body which is calculated by MC using the 20 levels radar echo data。

**Reply by ywszxdd** [reply]
the 3D radar body  data only has vertex、incides、colors，it is  many triangular primitive 。

**Reply by Zemledelec** [reply]
1) In this case, if you want some 3d geometry on the map, you have to create a custom scene and overwrite the frame method with your code, the same way you do in the example above, but don't use any `gl.clear` (renderer already do), the same way Planet does.  Check the example: http://openglobus.org/examples/customScene/customScene.html. Then, you could use Planet instance and your 3d scene together.

2) I strongly recommend, do not use `.set({'a_position': vertBuffer})` method from the shader program, I suggest using native webgl API.

3) wgs84.lonLatToCartesian input variable has LonLat type.

4) Cartesian coordinates in the Planet scale ellipsoid radius is equal to 6356752 meters, the near units you could use in your cartesian coordinates to show some in your scene, where Y is up, Z - is face vector, and X - is right. For more check this https://github.com/openglobus/openglobus/wiki/How-it-works#3d-projection-triangulation-and-geodetic-coordinates-ecef-translation

**Reply by Zemledelec** [reply]
P.S. You could start from displaying the point cloud, and after then will make 3d mesh.

**Reply by ywszxdd** [reply]
Thankyou very much , I'll study hard 。

**Reply by ywszxdd** [reply]
Hi@Zemledelec! I  created a custom scene and overwrite the frame method,then add the node to the globus.render,the triangle polygon colud show,but it still cannot change  when the globus change. this is my code , What's the problem?  P.S. “Then, you could use Planet instance and your 3d scene together.”mean?
```
    <script id="shader-vs" type="x-shader/x-vertes">
        attribute vec3 a_position;
        void main(){
          gl_Position = vec4(a_position,1.0);
        }
    </script>
    <script id="shader-fs" type="x-shader/x-fragment">
        precision mediump float;
        void main(){
          gl_FragColor = vec4(1,0,0.5,1);
        }
    </script>
    <script type="module">
        'use strict';
        import * as math from '../../src/og/math.js';
        import { RenderNode } from '../../src/og/scene/RenderNode.js';
        import { Program } from '../../src/og/webgl/Program.js';
        import { Globe } from '../../src/og/Globe.js';
        import { XYZ } from '../../src/og/layer/XYZ.js';
        import { GlobusTerrain } from '../../src/og/terrain/GlobusTerrain.js';
        import { wgs84 } from '../../src/og/ellipsoid/wgs84.js';
        import { LonLat } from '../../src/og/LonLat.js';
        class MyScene extends RenderNode {
            constructor() {
                super("MyScene");
                this.vericesBuffer = null;
            }
            init() {
                this.renderer.handler.addProgram(new Program("myShader", {
                    attributes: {
                        'a_position': 'vec3'
                    },
                    vertexShader: document.getElementById("shader-vs").innerHTML,
                    fragmentShader: document.getElementById("shader-fs").innerHTML
                }));
                var p1 = wgs84.lonLatToCartesian(new LonLat(0,0,1000)).normal();
                var p2 = wgs84.lonLatToCartesian(new LonLat(20,0,5000)).normal();
                var p3 = wgs84.lonLatToCartesian(new LonLat(10,20,1000)).normal();
                var positions = [
                    p1.x, p1.y, p1.z,
                    p2.x, p2.y, p2.z,
                    p3.x, p3.y, p3.z,
                ]
                this.vericesBuffer = this.renderer.handler.createArrayBuffer(new Float32Array(positions), 3, positions.length / 3);
            }
            frame() {
                var r = this.renderer;
                var sh = r.handler.programs.myShader;
                var p = sh._program;
                var gl = r.handler.gl;
                sh.activate();
                gl.bindBuffer(gl.ARRAY_BUFFER, this.vericesBuffer);
                gl.vertexAttribPointer(p.attributes.a_position, this.vericesBuffer.itemSize, gl.FLOAT, false, 0, 0);
                gl.drawArrays(gl.TRIANGLES, 0, 3)
            }
        }
        let osm = new XYZ("OpenStreetMap", {
            isBaseLayer: true,
            url: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
            visibility: true,
            attribution: 'Data @ OpenStreetMap contributors, ODbL'
        });
        let globus = new Globe({
            "target": "globus",
            "name": "Earth",
            "terrain": new GlobusTerrain(),
            "layers": [osm]
        });
        globus.renderer.addNodes([new MyScene()]);
    </script>
```

**Reply by Zemledelec** [reply]
Your shader works in plain screen coordinates. It means that your triangle works in 2D screen coordinates. If you want it displayed in 3D space of the planet scene, you also have to provide all view matrices the same way as in the example II sent before. 3D works the way that you pass x,y,z and convert the coordinates into 2d coordinates of your screen coordinates. Check (skim reading) this especially a part with matrices https://www3.ntu.edu.sg/home/ehchua/programming/opengl/CG_BasicsTheory.html

**Reply by ywszxdd** [reply]
Hi,I add the viewMatrix and the projectionMatrix , then the triangle can not display. What other matrix needs add?
<script id="shader-vs" type="x-shader/x-vertes">
        attribute vec3 a_position;
        uniform mat4 uMVMatrix;
        uniform mat4 uPMatrix;
        void main(){
            gl_Position = uPMatrix * uMVMatrix * vec4(a_position,1.0);
        }
    </script>
    <script id="shader-fs" type="x-shader/x-fragment">
        precision mediump float;
        void main(){
          gl_FragColor = vec4(1,0,0.5,1);
        }
    </script>
    <script type="module">
        'use strict';
        import * as math from '../../src/og/math.js';
        import { RenderNode } from '../../src/og/scene/RenderNode.js';
        import { Program } from '../../src/og/webgl/Program.js';
        import { Globe } from '../../src/og/Globe.js';
        import { XYZ } from '../../src/og/layer/XYZ.js';
        import { GlobusTerrain } from '../../src/og/terrain/GlobusTerrain.js';
        import { wgs84 } from '../../src/og/ellipsoid/wgs84.js';
        import { LonLat } from '../../src/og/LonLat.js';
        class MyScene extends RenderNode {
            constructor() {
                super("MyScene");
                this.vericesBuffer = null;
            }
            init() {
                this.renderer.handler.addProgram(new Program("myShader", {
                    uniforms: {
                        'uMVMatrix': 'mat4',
                        'uPMatrix': 'mat4'
                    },
                    attributes: {
                        'a_position': 'vec3'
                    },
                    vertexShader: document.getElementById("shader-vs").innerHTML,
                    fragmentShader: document.getElementById("shader-fs").innerHTML
                }));
                var p1 = wgs84.lonLatToCartesian(new LonLat(0,0,1000)).normal();
                var p2 = wgs84.lonLatToCartesian(new LonLat(20,0,5000)).normal();
                var p3 = wgs84.lonLatToCartesian(new LonLat(10,20,1000)).normal();
                var positions = [
                    p1.x, p1.y, p1.z,
                    p2.x, p2.y, p2.z,
                    p3.x, p3.y, p3.z,
                ]
                this.vericesBuffer = this.renderer.handler.createArrayBuffer(new Float32Array(positions), 3, positions.length / 3);
            }
            frame() {
                var r = this.renderer;
                var sh = r.handler.programs.myShader;
                var p = sh._program;
                var gl = r.handler.gl;
                sh.activate();

                var modelViewMat = r.activeCamera.getViewMatrix();
                var proMat = r.activeCamera.getProjectionMatrix();
                
                gl.uniformMatrix4fv(p.uniforms.uMVMatrix, false, modelViewMat);
                gl.uniformMatrix4fv(p.uniforms.uPMatrix, false, proMat);

                gl.bindBuffer(gl.ARRAY_BUFFER, this.vericesBuffer);
                gl.vertexAttribPointer(p.attributes.a_position, this.vericesBuffer.itemSize, gl.FLOAT, false, 0, 0);
                gl.drawArrays(gl.TRIANGLES, 0, 3)
            }
        }
        let osm = new XYZ("OpenStreetMap", {
            isBaseLayer: true,
            url: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
            visibility: true,
            attribution: 'Data @ OpenStreetMap contributors, ODbL'
        });
        let globus = new Globe({
            "target": "globus",
            "name": "Earth",
            "terrain": new GlobusTerrain(),
            "layers": [osm]
        });
        globus.renderer.addNodes([new MyScene()]);

**Reply by Zemledelec** [reply]
I strongly recommend to make examples and understand theory from here https://webglfundamentals.org/webgl/lessons/webgl-3d-perspective.html.

In your code, your took matrices right (the same as in the customScene example), but you have to make a math transformation from 3d coordinates to your 2d screen coordinates in a vertex shader, in the customScene example it looks like `gl_Position = uPMatrix * uMVMatrix * vec4(aVertexPosition, 1.0);` check it up:)

**Reply by ywszxdd** [reply]
I had make a math transformation from 3d coordinates to your 2d screen coordinates in a vertex shader,
<script id="shader-vs" type="x-shader/x-vertes"> attribute vec3 a_position; uniform mat4 uMVMatrix; uniform mat4 uPMatrix; void main(){ gl_Position = uPMatrix * uMVMatrix * vec4(a_position,1.0); } </script>
but it still could not display ,why....

**Reply by ywszxdd** [reply]
Plese check the next code is right?lon、lat、height covert to vertexBuffer
                var p1 = wgs84.lonLatToCartesian(new LonLat(0,0,1000)).normal();
                var p2 = wgs84.lonLatToCartesian(new LonLat(20,0,5000)).normal();
                var p3 = wgs84.lonLatToCartesian(new LonLat(10,20,1000)).normal();
                var positions = [
                    p1.x, p1.y, p1.z,
                    p2.x, p2.y, p2.z,
                    p3.x, p3.y, p3.z,
                ]
                this.vericesBuffer = this.renderer.handler.createArrayBuffer(new Float32Array(positions), 3, positions.length / 3);

**Reply by Zemledelec** [reply]
Try to remove the normalization of your cartesian coordinates.

**Reply by ywszxdd** [reply]
I found the question! "var p1 = wgs84.lonLatToCartesian(new LonLat(0,0,1000));"is correct。And the  height if is 5000,the triangle  is inside the sphere,so  I think it is not display。 
Thank you very much!!By the way,If I want to add my custom data as a entity not a scenenode , Can it be realized?

**Reply by Zemledelec** [reply]
Sure, but you need to know more about architecture of the entityCollections, and create some kind of "copy-paste" code:) On the other hand you can always use scenes, I use scenes all around on my side projects, and it works well!
